File: src/Application Framework/Entities/Users/UserSQL.vb
Imports System.Data.SqlClient
Imports FSM.Entity.Sys
Imports DocumentFormat.OpenXml.Drawing.Diagrams

Namespace Entity

    Namespace Users

        Public Class UserSQL

            Private _database As Sys.Database

            Public Sub New(ByVal databaseIn As Sys.Database)
                _database = databaseIn
            End Sub

#Region "Functions"

            Public Function Authenticate(ByVal UserName As String, ByVal Password As String) As Boolean
                _database.ClearParameter()
                _database.AddParameter("@UserName", UserName)
                _database.AddParameter("@Password", Entity.Sys.Helper.HashPassword(Password))
                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_Login_Mk1")
                If dt.Rows.Count > 0 Then
                    loggedInUser = [Get](Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("UserID")))
                    Return True
                Else
                    loggedInUser = Nothing
                    Return False
                End If
            End Function

            Public Function IsUserActive(Optional ByVal UserName As String = "",
                                         Optional ByVal UserID As Integer = 0,
                                         Optional ByVal SetProperty As Boolean = False) As Integer
                _database.ClearParameter()

                If Not String.IsNullOrEmpty(UserName) Then _database.AddParameter("@UserName", UserName)
                If UserID > 0 Then _database.AddParameter("@UserID", UserID)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_IsActive_Mk1")
                If dt.Rows.Count > 0 Then
                    If SetProperty Then loggedInUser = [Get](Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("UserID")), True)
                    Return True
                Else
                    Return False
                End If
            End Function

            Public Function Check_Unique_Username(ByVal username As String, ByVal ID As Integer) As Boolean
                Dim NumberOfUsers As Integer = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteScalar("SELECT COUNT(Userid) as 'NumberOfUsers' " &
                "FROM tbluser WHERE deleted = 0 AND username = '" & username & "' AND userid <> " & ID, False))

                If NumberOfUsers = 0 Then
                    Return True
                Else
                    Return False
                End If
            End Function

            Public Function GetAll() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_GetAll_Mk1")
                dt.TableName = Sys.Enums.TableNames.tblUser.ToString
                Return New DataView(dt)
            End Function

            Public Function GetAll_NoEngineers() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_GetAll_NoEngineers")
                dt.TableName = Sys.Enums.TableNames.tblUser.ToString
                Return New DataView(dt)
            End Function

            Public Function [Get](ByVal UserID As Integer, Optional ByVal passExpired As Boolean = False) As Entity.Users.User
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_Get_Mk1")

                If dt.Rows.Count > 0 Then
                    Dim userAndEngineer As New Entity.Users.User
                    userAndEngineer.SetUserID = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("UserID"))
                    userAndEngineer.SetFullname = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Fullname"))
                    userAndEngineer.SetUsername = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Username"))
                    userAndEngineer.SetPassword = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("Password"))
                    userAndEngineer.SetEnabled = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("Enabled"))
                    userAndEngineer.SetDeleted = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("Deleted"))
                    userAndEngineer.SetAdmin = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("Admin"))
                    userAndEngineer.SetSchedulerDayView = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("SchedulerDayView"))
                    If IsDBNull(dt.Rows(0).Item("DefaultSchedulerEngineerGroup")) Then
                        userAndEngineer.SetDefaultEngineerGroup = 0
                    Else
                        userAndEngineer.SetDefaultEngineerGroup = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("DefaultSchedulerEngineerGroup"))
                    End If
                    userAndEngineer.SetSchedulerTextSize = Entity.Sys.Helper.MakeIntegerValid(dt.Rows(0).Item("SchedulerTextSize"))
                    userAndEngineer.SetEmailAddress = Entity.Sys.Helper.MakeStringValid(dt.Rows(0).Item("EmailAddress"))
                    userAndEngineer.PasswordExpiryDate = Entity.Sys.Helper.MakeDateTimeValid(dt.Rows(0).Item("ExpiryDate"))
                    userAndEngineer.SetUpdateAtLogon = passExpired OrElse Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0)("UpdateAtLogon"))
                    userAndEngineer.SecurityUserModules = GetSecurityUserModules(UserID)
                    userAndEngineer.UserRegions = DB.User.UserRegions_Get(UserID)
                    userAndEngineer.Exists = True
                    Return userAndEngineer
                Else
                    Return Nothing
                End If
            End Function

            Public Function [GetAsync](ByVal UserID As Integer) As User
                Dim dt As New DataTable
                Dim command As New SqlCommand("User_Get_Mk1", New SqlConnection(_database.ServerConnectionString))
                Try
                    command.CommandType = CommandType.StoredProcedure
                    command.Parameters.AddWithValue("@UserID", UserID)
                    dt = _database.ExecuteCommand_DataTable(command)
                Catch ex As Exception
                Finally
                    command.Connection.Close()
                End Try

                If dt.Rows.Count > 0 Then
                    Dim user As New User With {
                        .SetUserID = Helper.MakeIntegerValid(dt.Rows(0).Item("UserID")),
                        .SetFullname = Helper.MakeStringValid(dt.Rows(0).Item("Fullname")),
                        .SetUsername = Helper.MakeStringValid(dt.Rows(0).Item("Username")),
                        .SetPassword = Helper.MakeStringValid(dt.Rows(0).Item("Password")),
                        .SetEnabled = Helper.MakeBooleanValid(dt.Rows(0).Item("Enabled")),
                        .SetDeleted = Helper.MakeBooleanValid(dt.Rows(0).Item("Deleted")),
                        .SetAdmin = Helper.MakeBooleanValid(dt.Rows(0).Item("Admin")),
                        .SetSchedulerDayView = Helper.MakeBooleanValid(dt.Rows(0).Item("SchedulerDayView")),
                        .SetDefaultEngineerGroup = Helper.MakeIntegerValid(dt.Rows(0).Item("DefaultSchedulerEngineerGroup")),
                        .SetSchedulerTextSize = Helper.MakeIntegerValid(dt.Rows(0).Item("SchedulerTextSize")),
                        .SetEmailAddress = Helper.MakeStringValid(dt.Rows(0).Item("EmailAddress")),
                        .PasswordExpiryDate = Helper.MakeDateTimeValid(dt.Rows(0).Item("ExpiryDate")),
                        .SetUpdateAtLogon = Helper.MakeBooleanValid(dt.Rows(0)("UpdateAtLogon"))
                    }
                    user.SecurityUserModules = GetSecurityUserModules(UserID)
                    user.UserRegions = DB.User.UserRegions_Get(UserID)
                    user.Exists = True
                    Return user
                Else
                    Return Nothing
                End If
            End Function

            Public Function Get_Passwords(ByVal UserID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("UserPassword_Get_ForUser_Mk1")
                Return New DataView(dt)
            End Function

            Public Function UpdatePassword(ByVal UserID As Integer, ByVal Password As String, Optional ByVal updateAtLogon As Boolean = False) As Boolean
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID)
                _database.AddParameter("@Password", Entity.Sys.Helper.HashPassword(Password))
                _database.AddParameter("@UpdateAtLogon", updateAtLogon)
                _database.AddParameter("@ExpiryDate", Now.AddMonths(3), True)

                Return CBool(_database.ExecuteSP_OBJECT("UserPassword_InsertUpdate_Mk1") = 1)
            End Function

            Public Function LastLogon(ByVal UserID As Integer) As String
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID)
                Try
                    Dim datetime As DateTime = Entity.Sys.Helper.MakeDateTimeValid(_database.ExecuteSP_OBJECT("User_LastLogon_Mk1"))
                    Return "Last logon : " & Format(datetime, "dd/MMM/yyyy HH:mm:ss") & "."
                Catch ex As Exception
                    Return "This is your first logon."
                End Try
            End Function

            Public Function Insert(ByVal userAndEngineer As Entity.Users.User) As Integer
                _database.ClearParameter()
                _database.AddParameter("@Fullname", userAndEngineer.Fullname)
                _database.AddParameter("@Username", userAndEngineer.Username)
                _database.AddParameter("@Password", Entity.Sys.Helper.HashPassword(userAndEngineer.Password))
                _database.AddParameter("@DefaultSchedulerEngineerGroup", userAndEngineer.DefaultEngineerGroup)
                _database.AddParameter("@EmailAddress", userAndEngineer.EmailAddress)
                _database.AddParameter("@Enabled", userAndEngineer.Enabled, True)
                _database.AddParameter("@Admin", userAndEngineer.Admin, True)
                _database.AddParameter("@SchedulerDayView", userAndEngineer.SchedulerDayView, True)

                Dim id As Integer = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("User_Insert_Mk1"))
                userAndEngineer.SetUserID = id
                SaveSecurityUserModules(userAndEngineer)
                Return id
            End Function

            Public Sub Update(ByVal userAndEngineer As Entity.Users.User)
                _database.ClearParameter()
                _database.AddParameter("@UserID", userAndEngineer.UserID)
                _database.AddParameter("@Fullname", userAndEngineer.Fullname)
                _database.AddParameter("@Username", userAndEngineer.Username)
                _database.AddParameter("@Password", Entity.Sys.Helper.HashPassword(userAndEngineer.Password))
                _database.AddParameter("@DefaultSchedulerEngineerGroup", userAndEngineer.DefaultEngineerGroup)
                _database.AddParameter("@EmailAddress", userAndEngineer.EmailAddress)
                _database.AddParameter("@Enabled", userAndEngineer.Enabled, True)
                _database.AddParameter("@Admin", userAndEngineer.Admin, True)
                _database.AddParameter("@SchedulerDayView", userAndEngineer.SchedulerDayView, True)

                _database.ExecuteSP_OBJECT("User_Update_Mk1")
                SaveSecurityUserModules(userAndEngineer)
            End Sub

            Public Sub Delete(ByVal UserID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID)
                _database.ExecuteSP_OBJECT("User_Delete_Mk1")
            End Sub

            Public Function User_Search(ByVal criteria As String) As DataView
                _database.ClearParameter()
                _database.AddParameter("@SearchString", criteria, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("User_SearchList_Mk1")
                dt.TableName = Entity.Sys.Enums.TableNames.tblUser.ToString
                Return New DataView(dt)
            End Function

            Public Sub User_Update_TextSize(ByVal UserID As Integer, ByVal TextSize As Integer)
                _database.ClearParameter()
                _database.ExecuteScalar("UPDATE tblUser Set SchedulerTextSize = " & TextSize & " Where UserID = " & UserID)

            End Sub

#End Region

#Region "Security"

            Public Function GetSecurityUserModulesDefaults() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("SecurityUserModulesDefault_Get")
                dt.TableName = Sys.Enums.TableNames.tblUser.ToString
                Return New DataView(dt)
            End Function

            Public Function GetUserRegions(ByVal userId As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@UserID", userId, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("UserRegions_Get_ForUserID")
                dt.TableName = Sys.Enums.TableNames.tblRegion.ToString
                Return New DataView(dt)
            End Function

            Public Function GetUserRegionsDefaults() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("UserRegions_Get_Defaults")
                dt.TableName = Sys.Enums.TableNames.tblRegion.ToString
                Return New DataView(dt)
            End Function

            Public Function GetSecurityUserModules(ByVal userID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@UserID", userID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("SecurityUserModules_Get")
                dt.TableName = Entity.Sys.Enums.TableNames.tblSecurityUserModules.ToString()
                Return New DataView(dt)
            End Function

            Private Function SaveSecurityUserModules(ByVal u As User) As User
                If Not u.SecurityUserModules Is Nothing Then
                    Dim t As DataTable = Nothing
                    t = u.SecurityUserModules.Table

                    Dim diff As DataTable = t.GetChanges()
                    If diff IsNot Nothing Then
                        Dim changes As String = String.Empty
                        Dim addAnd As Boolean = False
                        For Each dr As DataRow In diff.Rows
                            If addAnd Then changes += " and "
                            changes += Sys.Helper.MakeStringValid(dr("ModuleName")) & " set to " & Sys.Helper.MakeStringValid(dr("AccessPermitted"))
                            addAnd = True
                        Next
                        If Not String.IsNullOrEmpty(changes) Then UserAccessAudit_Insert(u.UserID, changes)
                    End If

                    For Each r As DataRow In t.Rows
                        _database.ClearParameter()
                        _database.AddParameter("@UserID", u.UserID, True)
                        _database.AddParameter("@SystemModuleID", CInt(r("SystemModuleID")), True)
                        _database.AddParameter("@AccessPermitted", CBool(r("AccessPermitted")), True)
                        If CInt(r("UserModuleID")) = 0 Then
                            _database.ExecuteSP_NO_Return("SecurityUserModules_Insert")
                        Else
                            _database.ExecuteSP_NO_Return("SecurityUserModules_Update")
                        End If
                    Next

                    u.SecurityUserModules = GetSecurityUserModules(u.UserID)
                End If
                Return u
            End Function

            Public Sub InsertNewSecurityUserModules(ByVal UserID As Integer, ByVal SystemModuleID As Integer, ByVal AccessPermitted As Boolean)
                _database.ClearParameter()
                _database.AddParameter("@UserID", UserID, True)
                _database.AddParameter("@SystemModuleID", SystemModuleID, True)
                _database.AddParameter("@AccessPermitted", AccessPermitted, True)
                _database.ExecuteSP_NO_Return("SecurityUserModules_Insert")
            End Sub

            Public Function UserRegions_Get(ByVal userID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@UserID", userID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("UserRegions_Get")
                Return New DataView(dt)
            End Function

            Public Sub UserRegions_Insert(ByVal userID As Integer, ByVal managerID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@UserID", userID, True)
                _database.AddParameter("@RegionID", managerID, True)
                _database.ExecuteSP_OBJECT("UserRegion_Insert")
            End Sub

            Public Sub UserRegions_Delete(ByVal userID As Integer)
                _database.ClearParameter()
                DB.ExecuteScalar("DELETE FROM tblUserRegion Where UserID = " & userID)
            End Sub

            Public Sub [UserAccessAudit_Insert](ByVal userId As Integer, ByVal actionChange As String)
                _database.ClearParameter()

                _database.AddParameter("@UserID", userId, True)
                _database.AddParameter("@ActionChange", actionChange, True)
                _database.AddParameter("@ActionDateTime", DateTime.Now, True)
                _database.AddParameter("@AuthUserID", loggedInUser.UserID, True)

                _database.ExecuteSP_NO_Return("UserAccessAudit_Insert")
            End Sub

#End Region

        End Class

    End Namespace

End Namespace

File: apollo/apollo.fsm/src/Application Framework/Entities/Users/UserSQL.cs
using FSM.Entity.Sys;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Data;
using System.Data.SqlClient;

namespace FSM.Entity
{
    namespace Users
    {
        public class UserSQL
        {
            private Database _database;

            public UserSQL(Database databaseIn)
            {
                _database = databaseIn;
            }

            

            public bool Authenticate(string UserName, string Password)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserName", UserName);
                _database.AddParameter("@Password", Helper.HashPassword(Password));
                var dt = _database.ExecuteSP_DataTable("User_Login_Mk1");
                if (dt.Rows.Count > 0)
                {
                    App.loggedInUser = Get(Helper.MakeIntegerValid(dt.Rows[0]["UserID"]));
                    return true;
                }
                else
                {
                    App.loggedInUser = null;
                    return false;
                }
            }

            public int IsUserActive(string UserName = "", int UserID = 0, bool SetProperty = false)
            {
                _database.ClearParameter();
                if (!string.IsNullOrEmpty(UserName))
                    _database.AddParameter("@UserName", UserName);
                if (UserID > 0)
                    _database.AddParameter("@UserID", UserID);
                var dt = _database.ExecuteSP_DataTable("User_IsActive_Mk1");
                if (dt.Rows.Count > 0)
                {
                    if (SetProperty)
                        App.loggedInUser = Get(Helper.MakeIntegerValid(dt.Rows[0]["UserID"]), true);
                    return Conversions.ToInteger(true);
                }
                else
                {
                    return Conversions.ToInteger(false);
                }
            }

            public bool Check_Unique_Username(string username, int ID)
            {
                int NumberOfUsers = Helper.MakeIntegerValid(_database.ExecuteScalar("SELECT COUNT(Userid) as 'NumberOfUsers' " + "FROM tbluser WHERE deleted = 0 AND username = '" + username + "' AND userid <> " + ID, false));
                if (NumberOfUsers == 0)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }

            public DataView GetAll()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("User_GetAll_Mk1");
                dt.TableName = Enums.TableNames.tblUser.ToString();
                return new DataView(dt);
            }

            public DataView GetAll_NoEngineers()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("User_GetAll_NoEngineers");
                dt.TableName = Enums.TableNames.tblUser.ToString();
                return new DataView(dt);
            }

            public User Get(int UserID, bool passExpired = false)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID);
                var dt = _database.ExecuteSP_DataTable("User_Get_Mk1");
                if (dt.Rows.Count > 0)
                {
                    var userAndEngineer = new User();
                    userAndEngineer.SetUserID = Helper.MakeIntegerValid(dt.Rows[0]["UserID"]);
                    userAndEngineer.SetFullname = Helper.MakeStringValid(dt.Rows[0]["Fullname"]);
                    userAndEngineer.SetUsername = Helper.MakeStringValid(dt.Rows[0]["Username"]);
                    userAndEngineer.SetPassword = Helper.MakeStringValid(dt.Rows[0]["Password"]);
                    userAndEngineer.SetEnabled = Helper.MakeBooleanValid(dt.Rows[0]["Enabled"]);
                    userAndEngineer.SetDeleted = Helper.MakeBooleanValid(dt.Rows[0]["Deleted"]);
                    userAndEngineer.SetAdmin = Helper.MakeBooleanValid(dt.Rows[0]["Admin"]);
                    userAndEngineer.SetSchedulerDayView = Helper.MakeBooleanValid(dt.Rows[0]["SchedulerDayView"]);
                    if (Information.IsDBNull(dt.Rows[0]["DefaultSchedulerEngineerGroup"]))
                    {
                        userAndEngineer.SetDefaultEngineerGroup = 0;
                    }
                    else
                    {
                        userAndEngineer.SetDefaultEngineerGroup = Helper.MakeIntegerValid(dt.Rows[0]["DefaultSchedulerEngineerGroup"]);
                    }

                    userAndEngineer.SetSchedulerTextSize = Helper.MakeIntegerValid(dt.Rows[0]["SchedulerTextSize"]);
                    userAndEngineer.SetEmailAddress = Helper.MakeStringValid(dt.Rows[0]["EmailAddress"]);
                    userAndEngineer.PasswordExpiryDate = Helper.MakeDateTimeValid(dt.Rows[0]["ExpiryDate"]);
                    userAndEngineer.SetUpdateAtLogon = passExpired || Helper.MakeBooleanValid(dt.Rows[0]["UpdateAtLogon"]);
                    userAndEngineer.SecurityUserModules = GetSecurityUserModules(UserID);
                    userAndEngineer.UserRegions = App.DB.User.UserRegions_Get(UserID);
                    userAndEngineer.Exists = true;
                    return userAndEngineer;
                }
                else
                {
                    return null;
                }
            }

            public User GetAsync(int UserID)
            {
                var dt = new DataTable();
                var command = new SqlCommand("User_Get_Mk1", new SqlConnection(_database.ServerConnectionString));
                try
                {
                    command.CommandType = CommandType.StoredProcedure;
                    command.Parameters.AddWithValue("@UserID", UserID);
                    dt = _database.ExecuteCommand_DataTable(command);
                }
                catch (Exception ex)
                {
                }
                finally
                {
                    command.Connection.Close();
                }

                if (dt.Rows.Count > 0)
                {
                    var user = new User()
                    {
                        SetUserID = Helper.MakeIntegerValid(dt.Rows[0]["UserID"]),
                        SetFullname = Helper.MakeStringValid(dt.Rows[0]["Fullname"]),
                        SetUsername = Helper.MakeStringValid(dt.Rows[0]["Username"]),
                        SetPassword = Helper.MakeStringValid(dt.Rows[0]["Password"]),
                        SetEnabled = Helper.MakeBooleanValid(dt.Rows[0]["Enabled"]),
                        SetDeleted = Helper.MakeBooleanValid(dt.Rows[0]["Deleted"]),
                        SetAdmin = Helper.MakeBooleanValid(dt.Rows[0]["Admin"]),
                        SetSchedulerDayView = Helper.MakeBooleanValid(dt.Rows[0]["SchedulerDayView"]),
                        SetDefaultEngineerGroup = Helper.MakeIntegerValid(dt.Rows[0]["DefaultSchedulerEngineerGroup"]),
                        SetSchedulerTextSize = Helper.MakeIntegerValid(dt.Rows[0]["SchedulerTextSize"]),
                        SetEmailAddress = Helper.MakeStringValid(dt.Rows[0]["EmailAddress"]),
                        PasswordExpiryDate = Helper.MakeDateTimeValid(dt.Rows[0]["ExpiryDate"]),
                        SetUpdateAtLogon = Helper.MakeBooleanValid(dt.Rows[0]["UpdateAtLogon"])
                    };
                    user.SecurityUserModules = GetSecurityUserModules(UserID);
                    user.UserRegions = App.DB.User.UserRegions_Get(UserID);
                    user.Exists = true;
                    return user;
                }
                else
                {
                    return null;
                }
            }

            public DataView Get_Passwords(int UserID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID);
                var dt = _database.ExecuteSP_DataTable("UserPassword_Get_ForUser_Mk1");
                return new DataView(dt);
            }

            public bool UpdatePassword(int UserID, string Password, bool updateAtLogon = false)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID);
                _database.AddParameter("@Password", Helper.HashPassword(Password));
                _database.AddParameter("@UpdateAtLogon", updateAtLogon);
                _database.AddParameter("@ExpiryDate", DateAndTime.Now.AddMonths(3), true);
                return Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(_database.ExecuteSP_OBJECT("UserPassword_InsertUpdate_Mk1"), 1, false));
            }

            public string LastLogon(int UserID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID);
                try
                {
                    var datetime = Helper.MakeDateTimeValid(_database.ExecuteSP_OBJECT("User_LastLogon_Mk1"));
                    return "Last logon : " + Strings.Format(datetime, "dd/MMM/yyyy HH:mm:ss") + ".";
                }
                catch (Exception ex)
                {
                    return "This is your first logon.";
                }
            }

            public int Insert(User userAndEngineer)
            {
                _database.ClearParameter();
                _database.AddParameter("@Fullname", userAndEngineer.Fullname);
                _database.AddParameter("@Username", userAndEngineer.Username);
                _database.AddParameter("@Password", Helper.HashPassword(userAndEngineer.Password));
                _database.AddParameter("@DefaultSchedulerEngineerGroup", userAndEngineer.DefaultEngineerGroup);
                _database.AddParameter("@EmailAddress", userAndEngineer.EmailAddress);
                _database.AddParameter("@Enabled", userAndEngineer.Enabled, true);
                _database.AddParameter("@Admin", userAndEngineer.Admin, true);
                _database.AddParameter("@SchedulerDayView", userAndEngineer.SchedulerDayView, true);
                int id = Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("User_Insert_Mk1"));
                userAndEngineer.SetUserID = id;
                SaveSecurityUserModules(userAndEngineer);
                return id;
            }

            public void Update(User userAndEngineer)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userAndEngineer.UserID);
                _database.AddParameter("@Fullname", userAndEngineer.Fullname);
                _database.AddParameter("@Username", userAndEngineer.Username);
                _database.AddParameter("@Password", Helper.HashPassword(userAndEngineer.Password));
                _database.AddParameter("@DefaultSchedulerEngineerGroup", userAndEngineer.DefaultEngineerGroup);
                _database.AddParameter("@EmailAddress", userAndEngineer.EmailAddress);
                _database.AddParameter("@Enabled", userAndEngineer.Enabled, true);
                _database.AddParameter("@Admin", userAndEngineer.Admin, true);
                _database.AddParameter("@SchedulerDayView", userAndEngineer.SchedulerDayView, true);
                _database.ExecuteSP_OBJECT("User_Update_Mk1");
                SaveSecurityUserModules(userAndEngineer);
            }

            public void Delete(int UserID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID);
                _database.ExecuteSP_OBJECT("User_Delete_Mk1");
            }

            public DataView User_Search(string criteria)
            {
                _database.ClearParameter();
                _database.AddParameter("@SearchString", criteria, true);
                var dt = _database.ExecuteSP_DataTable("User_SearchList_Mk1");
                dt.TableName = Enums.TableNames.tblUser.ToString();
                return new DataView(dt);
            }

            public void User_Update_TextSize(int UserID, int TextSize)
            {
                _database.ClearParameter();
                _database.ExecuteScalar("UPDATE tblUser Set SchedulerTextSize = " + TextSize + " Where UserID = " + UserID);
            }

            
            

            public DataView GetSecurityUserModulesDefaults()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("SecurityUserModulesDefault_Get");
                dt.TableName = Enums.TableNames.tblUser.ToString();
                return new DataView(dt);
            }

            public DataView GetUserRegions(int userId)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userId, true);
                var dt = _database.ExecuteSP_DataTable("UserRegions_Get_ForUserID");
                dt.TableName = Enums.TableNames.tblRegion.ToString();
                return new DataView(dt);
            }

            public DataView GetUserRegionsDefaults()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("UserRegions_Get_Defaults");
                dt.TableName = Enums.TableNames.tblRegion.ToString();
                return new DataView(dt);
            }

            public DataView GetSecurityUserModules(int userID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userID, true);
                var dt = _database.ExecuteSP_DataTable("SecurityUserModules_Get");
                dt.TableName = Enums.TableNames.tblSecurityUserModules.ToString();
                return new DataView(dt);
            }

            private User SaveSecurityUserModules(User u)
            {
                if (u.SecurityUserModules is object)
                {
                    DataTable t = null;
                    t = u.SecurityUserModules.Table;
                    var diff = t.GetChanges();
                    if (diff is object)
                    {
                        string changes = string.Empty;
                        bool addAnd = false;
                        foreach (DataRow dr in diff.Rows)
                        {
                            if (addAnd)
                                changes += " and ";
                            changes += Helper.MakeStringValid(dr["ModuleName"]) + " set to " + Helper.MakeStringValid(dr["AccessPermitted"]);
                            addAnd = true;
                        }

                        if (!string.IsNullOrEmpty(changes))
                            UserAccessAudit_Insert(u.UserID, changes);
                    }

                    foreach (DataRow r in t.Rows)
                    {
                        _database.ClearParameter();
                        _database.AddParameter("@UserID", u.UserID, true);
                        _database.AddParameter("@SystemModuleID", Conversions.ToInteger(r["SystemModuleID"]), true);
                        _database.AddParameter("@AccessPermitted", Conversions.ToBoolean(r["AccessPermitted"]), true);
                        if (Conversions.ToInteger(r["UserModuleID"]) == 0)
                        {
                            _database.ExecuteSP_NO_Return("SecurityUserModules_Insert");
                        }
                        else
                        {
                            _database.ExecuteSP_NO_Return("SecurityUserModules_Update");
                        }
                    }

                    u.SecurityUserModules = GetSecurityUserModules(u.UserID);
                }

                return u;
            }

            public void InsertNewSecurityUserModules(int UserID, int SystemModuleID, bool AccessPermitted)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", UserID, true);
                _database.AddParameter("@SystemModuleID", SystemModuleID, true);
                _database.AddParameter("@AccessPermitted", AccessPermitted, true);
                _database.ExecuteSP_NO_Return("SecurityUserModules_Insert");
            }

            public DataView UserRegions_Get(int userID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userID, true);
                var dt = _database.ExecuteSP_DataTable("UserRegions_Get");
                return new DataView(dt);
            }

            public void UserRegions_Insert(int userID, int managerID)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userID, true);
                _database.AddParameter("@RegionID", managerID, true);
                _database.ExecuteSP_OBJECT("UserRegion_Insert");
            }

            public void UserRegions_Delete(int userID)
            {
                _database.ClearParameter();
                App.DB.ExecuteScalar("DELETE FROM tblUserRegion Where UserID = " + userID);
            }

            public void UserAccessAudit_Insert(int userId, string actionChange)
            {
                _database.ClearParameter();
                _database.AddParameter("@UserID", userId, true);
                _database.AddParameter("@ActionChange", actionChange, true);
                _database.AddParameter("@ActionDateTime", DateTime.Now, true);
                _database.AddParameter("@AuthUserID", App.loggedInUser.UserID, true);
                _database.ExecuteSP_NO_Return("UserAccessAudit_Insert");
            }

            
        }
    }
}